packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1005,16 +1085,20 @@ class ContainerLayer extends Layer {
 
   @override
   void detach() {
+    assert(!_debugMutationsLocked);
     super.detach();
     Layer? child = firstChild;
     while (child != null) {
       child.detach();
       child = child.nextSibling;
     }
+    // Children fired them already in child.detach().
+    _fireCompositionCallbacks(false);
```

Comment:
Why do we need to fire on detach?

Reply:
Detach indicates we're _not_ going to get drawn again.  If we don't fire here, we risk telling clients that we rendered something and never updating when we got dropped.

Attach shouldn't be needed because if you get composed after attach buildScene will be called.

I'll update the comment.

Reply:
Maybe also update the doc comment explaining when the callback will be called. From the current description my understanding was it would only happen when the layer is still part of the scene, not when it is going away...

Reply:
Done